compiler/rustc_mir_build/src/build/expr/as_rvalue.rs

```diff
@@ -174,7 +175,36 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
 
                 block.and(Rvalue::Aggregate(box AggregateKind::Tuple, fields))
             }
-            ExprKind::Closure { closure_id, substs, upvars, movability } => {
+            ExprKind::Closure { closure_id, substs, upvars, movability, fake_reads } => {
+                // Convert the closure fake reads, if any, from `ExprRef` to mir `Place`
+                // and push the fake reads.
+                // This must come before creating the operands. This is required in case
+                // there is a fake read and a borrow of the same path, since otherwise the
+                // fake read might interfere with the borrow. Consider an example like this
+                // one:
+                // ```
+                // let mut x = 0;
+                // let c = || {
+                //     &mut x; // mutable borrow of `x`
+                //     match x { _ => () } // fake read of `x`
+                // };
+                // ```
+
+                for (thir_place, cause) in fake_reads.into_iter() {
+                    let fake_read_upvar = this.hir.mirror(thir_place);
+                    let place_builder =
+                        unpack!(block = this.as_place_builder(block, fake_read_upvar));
+
+                    if let Ok(place_builder_resolved) =
+                        place_builder.try_upvars_resolved(this.hir.tcx(), this.hir.typeck_results())
+                    {
+                        let mir_place = place_builder_resolved
+                            .clone()
+                            .into_place(this.hir.tcx(), this.hir.typeck_results());
+                        this.cfg.push_fake_read(block, source_info, cause, mir_place);
```

Comment:
@roxelo so the source_info here is the span of the expression currently being processed, which in this case is the closure itself. 

In the test case where the diagnostics are currently affected, the fake reads conflicts with the place that was moved earlier and for diagnostics we are providing the entire closure.

To get proper diagnostics we want to have the correct span, so maybe we add a `diag_expr` similar to other ExprUseVisitor::Delegate methods and store that in typeck results.

Reply:
There is sort of another question at the back of my mind, which is should the fake reads have caused this conflict?

On the surface I think it makes sense for them to do so, because we shouldn't be able to even fake read a place that has already been moved, but would be nice if we can confirm this is valid/acceptable behavior

Reply:
for eg:

```rust
fn closure() {
    let mut tup = (U, U, U); // 8
    let c1 = || { // 9
        let (ref _x0, _x1, _) = tup; // 10
    }; // 11
    let c2 =  || { // 12
        //~^ ERROR use of moved value
        let (ref mut _x0, _, _x2) = tup;
    }; // 15
    drop(c1);
}
```
mir is:

```
        FakeRead(ForLet, _1);            // scope 0 at borrowck-move-ref-pattern.rs:8:9: 8:16
        StorageLive(_5);                 // scope 1 at borrowck-move-ref-pattern.rs:9:9: 9:11
        FakeRead(ForLet, _1);            // scope 1 at borrowck-move-ref-pattern.rs:9:14: 11:6
        _5 = [closure@borrowck-move-ref-pattern.rs:9:14: 11:6] { tup: move _1 }; // scope 1 at borrowck-move-ref-pattern.rs:9:14: 11:6
                                         // closure
                                         // + def_id: DefId(0:7 ~ borrowck_move_ref_pattern[317d]::closure::{closure#0})
                                         // + substs: [
                                         //     i32,
                                         //     extern "rust-call" fn(()),
                                         //     ((U, U, U),),
                                         // ]
        FakeRead(ForLet, _5);            // scope 1 at borrowck-move-ref-pattern.rs:9:9: 9:11
        StorageLive(_6);                 // scope 2 at borrowck-move-ref-pattern.rs:12:9: 12:11
        FakeRead(ForLet, _1);            // scope 2 at borrowck-move-ref-pattern.rs:12:15: 15:6
        _6 = [closure@borrowck-move-ref-pattern.rs:12:15: 15:6] { tup: move _1 }; // scope 2 at borrowck-move-ref-pattern.rs:12:15: 15:6
```